[MSE][GStreamer] Make fake preroll asynchronous

eocanha opened the pull request:
#### ff888612263c267aeb43dfc51e37af3264dd92f1
<pre>
[MSE][GStreamer] Make fake preroll asynchronous
<a href="https://bugs.webkit.org/show_bug.cgi?id=295289">https://bugs.webkit.org/show_bug.cgi?id=295289</a>

Reviewed by Alicia Boya Garcia.

With the current implementation the whole seek flow as executed in a
single main loop cycle. As a result, the JS app had no chance to spot
that video.seeking attribute is ever set to true. Also some apps listen
to video.onseeking and expects that video.seeking attribute to be true
in such case. That wasn&apos;t a valid assumption.

Also, the behaviour isn&apos;t the same in &lt;video&gt; elements that are
audio-only as in &lt;audio&gt; elements.

This happens on the Amlogic platform, that doesn&apos;t have asynchronous
state changes on its audio sink. Spotify is broken there because of this
problem.

See:
<a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1232">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1232</a>
<a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1527">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1527</a>

This patch calls didPreroll() also for &lt;video&gt; elements that are
audio-only. Also makes that call async by putting it on the HTML media
element task queue to make sure it is executed after dispatching
&apos;seeking&apos; event to JS so the app has a chance to notice that
HTMLmedia.seeking attribute is &apos;true&apos;.

Original author: Andrzej Surdej &lt;101130014+asurdej-comcast@users.noreply.github.com&gt;

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::doSeek): Find the actual final audio sink. Call didPreroll() asynchronously for any media element having only audio.

Canonical link: <a href="https://commits.webkit.org/298239@main">https://commits.webkit.org/298239@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f6ceb031bcaefefa2ed091323e8b85012a35c00

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114807 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34552 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25016 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120969 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65496 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116696 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35181 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43110 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/87261 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117755 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28032 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103093 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67652 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27217 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21215 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64637 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97416 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/21328 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124170 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41811 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/31236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/96067 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42189 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99282 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95852 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24398 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/41037 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/18877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37866 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41692 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47206 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41254 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44569 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42998 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc reviewed (approved):
LGTM.

I find it concerning that amlogic's audio sink is `async=false` and I wouldn't be surprised if that caused A/V synchronization issues, but bearing that in mind, the WebKit changes make sense.

eocanha committed:
[MSE][GStreamer] Make fake preroll asynchronous
https://bugs.webkit.org/show_bug.cgi?id=295289

Reviewed by Alicia Boya Garcia.

With the current implementation the whole seek flow as executed in a
single main loop cycle. As a result, the JS app had no chance to spot
that video.seeking attribute is ever set to true. Also some apps listen
to video.onseeking and expects that video.seeking attribute to be true
in such case. That wasn't a valid assumption.

Also, the behaviour isn't the same in <video> elements that are
audio-only as in <audio> elements.

This happens on the Amlogic platform, that doesn't have asynchronous
state changes on its audio sink. Spotify is broken there because of this
problem.

See:
https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1232
https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1527

This patch calls didPreroll() also for <video> elements that are
audio-only. Also makes that call async by putting it on the HTML media
element task queue to make sure it is executed after dispatching
'seeking' event to JS so the app has a chance to notice that
HTMLmedia.seeking attribute is 'true'.

Original author: Andrzej Surdej <101130014+asurdej-comcast@users.noreply.github.com>

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::doSeek): Find the actual final audio sink. Call didPreroll() asynchronously for any media element having only audio.

Canonical link: https://commits.webkit.org/298239@main
